refactor(linter): update tsgolint payload

camchenry opened the pull request:
- oxc side of https://github.com/oxc-project/tsgolint/pull/182

Updates the payload that `oxlint` sends to `tsgolint` to match the new format. The new structure is based around sets of rules and configurations first, and then specifying the files those configs act upon. That looks like something like this (output from running `oxlint --type-aware apps/oxlint/fixtures`)

```json
{
  "configs": [
    {
      "file_paths": [
        "/Users/<user>/workspace/oxc/apps/oxlint/fixtures/tsgolint/use-unknown-in-catch-callback-variable/index.ts",
        "/Users/<user>/workspace/oxc/apps/oxlint/fixtures/tsgolint/prefer-reduce-type-parameter/index.ts",
        "/Users/<user>/workspace/oxc/apps/oxlint/fixtures/tsgolint/return-await/index.ts",
        "/Users/<user>/workspace/oxc/apps/oxlint/fixtures/tsgolint/non-tsgolint.ts",
        "..."
        "/Users/<user>/workspace/oxc/apps/oxlint/fixtures/tsgolint/no-unnecessary-template-expression/index.ts",
        "/Users/<user>/workspace/oxc/apps/oxlint/fixtures/tsgolint/no-unsafe-type-assertion/index.ts",
        "/Users/<user>/workspace/oxc/apps/oxlint/fixtures/tsgolint/no-unsafe-member-access/index.ts",
        "/Users/<user>/workspace/oxc/apps/oxlint/fixtures/tsgolint/no-unsafe-call/index.ts",
        "/Users/<user>/workspace/oxc/apps/oxlint/fixtures/tsgolint/no-unnecessary-boolean-literal-compare/index.ts"
      ],
      "rules": [
        { "name": "await-thenable" },
        { "name": "no-floating-promises" },
        { "name": "no-array-delete" },
        { "name": "no-base-to-string" },
        { "name": "no-confusing-void-expression" },
        { "name": "no-duplicate-type-constituents" },
        { "...": "..."},
        { "name": "restrict-plus-operands" },
        { "name": "restrict-template-expressions" },
        { "name": "return-await" },
        { "name": "switch-exhaustiveness-check" },
        { "name": "unbound-method" },
        { "name": "use-unknown-in-catch-callback-variable" }
      ]
    },
    {
      "file_paths": [
        "/Users/<user>/workspace/oxc/apps/oxlint/fixtures/tsgolint/no-floating-promises/src/index.ts",
        "/Users/<user>/workspace/oxc/apps/oxlint/fixtures/tsgolint/no-floating-promises/src/overrides.ts"
      ],
      "rules": [{ "name": "no-floating-promises" }]
    }
  ]
}
```

camchenry commented:
* **#13547** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/13547?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/13547?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=13547).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

camchenry commented:
@Boshen @camc314  This will be needed for the next version of oxlint since this tsgolint update merged which made the configs incompatible: https://github.com/oxc-project/tsgolint/pull/182

camc314 commented:
> @Boshen @camc314  This will be needed for the next version of oxlint since this tsgolint update merged which made the configs incompatible: https://github.com/oxc-project/tsgolint/pull/182

i changed the pr to make it backwards compatible. i'm going to merge this tomorrow

camchenry commented:
Ah okay, sorry I missed that. All good then!

camchenry committed:
refactor(linter): update tsgolint payload (#13547)

- oxc side of https://github.com/oxc-project/tsgolint/pull/182

Updates the payload that `oxlint` sends to `tsgolint` to match the new format. The new structure is based around sets of rules and configurations first, and then specifying the files those configs act upon. That looks like something like this (output from running `oxlint --type-aware apps/oxlint/fixtures`)

```json
{
  "configs": [
    {
      "file_paths": [
        "/Users/<user>/workspace/oxc/apps/oxlint/fixtures/tsgolint/use-unknown-in-catch-callback-variable/index.ts",
        "/Users/<user>/workspace/oxc/apps/oxlint/fixtures/tsgolint/prefer-reduce-type-parameter/index.ts",
        "/Users/<user>/workspace/oxc/apps/oxlint/fixtures/tsgolint/return-await/index.ts",
        "/Users/<user>/workspace/oxc/apps/oxlint/fixtures/tsgolint/non-tsgolint.ts",
        "..."
        "/Users/<user>/workspace/oxc/apps/oxlint/fixtures/tsgolint/no-unnecessary-template-expression/index.ts",
        "/Users/<user>/workspace/oxc/apps/oxlint/fixtures/tsgolint/no-unsafe-type-assertion/index.ts",
        "/Users/<user>/workspace/oxc/apps/oxlint/fixtures/tsgolint/no-unsafe-member-access/index.ts",
        "/Users/<user>/workspace/oxc/apps/oxlint/fixtures/tsgolint/no-unsafe-call/index.ts",
        "/Users/<user>/workspace/oxc/apps/oxlint/fixtures/tsgolint/no-unnecessary-boolean-literal-compare/index.ts"
      ],
      "rules": [
        { "name": "await-thenable" },
        { "name": "no-floating-promises" },
        { "name": "no-array-delete" },
        { "name": "no-base-to-string" },
        { "name": "no-confusing-void-expression" },
        { "name": "no-duplicate-type-constituents" },
        { "...": "..."},
        { "name": "restrict-plus-operands" },
        { "name": "restrict-template-expressions" },
        { "name": "return-await" },
        { "name": "switch-exhaustiveness-check" },
        { "name": "unbound-method" },
        { "name": "use-unknown-in-catch-callback-variable" }
      ]
    },
    {
      "file_paths": [
        "/Users/<user>/workspace/oxc/apps/oxlint/fixtures/tsgolint/no-floating-promises/src/index.ts",
        "/Users/<user>/workspace/oxc/apps/oxlint/fixtures/tsgolint/no-floating-promises/src/overrides.ts"
      ],
      "rules": [{ "name": "no-floating-promises" }]
    }
  ]
}
```
